src/moonlink/src/union_read/read_state_manager.rs

```diff
@@ -133,18 +137,50 @@ impl ReadStateManager {
         }
     }
 
+    #[inline]
+    fn validate_lsn_ordering(
+        snapshot_lsn: u64,
+        commit_lsn: u64,
+        replication_lsn: u64,
+    ) -> Result<()> {
+        // validate right ordering lsn: snapshot_lsn<=commit_lsn<=replication_lsn
+        if snapshot_lsn != NO_SNAPSHOT_LSN
+            && commit_lsn != NO_COMMIT_LSN
+            && snapshot_lsn > commit_lsn
+        {
+            return Err(Error::read_validation_error(format!(
+                "snapshot_lsn > commit_lsn: {} > {}",
+                snapshot_lsn, commit_lsn
+            )));
+        }
+        if commit_lsn > replication_lsn {
```

Comment:
What's the purpose of retry?

Reply:
The purpose of retries is to synchronize state from the next function `wait_for_relevant_lsn_change`, because the events are received from independent channels, which leads to asynchronous problems. You can reproduce the issue in my test case by setting const MAX_READ_SNAPSHOT_RETRIES: u8 = 0 to observe the problems.
In my case, if MAX_READ_SNAPSHOT_RETRIES: u8 = 0, I see 36 cases which exist the problem: replication_lsn<= commit_lsn. Causing a failure situation

Reply:
That's because I see MoonLink handles many background events, for example:
CdcEvent::PrimaryKeepAlive
CdcEvent::StreamCommit
CdcEvent::Commit
... etc.

Each Message Passing channel can be used different Event or anywhere. To boost performance and make the code easier to scale, I see your team has broken it into many channels and run them asynchronously

Reply:
Let me check. What about table_snapshot_watch_receiver? It is also an independent receiver and completely operates asynchronously.

Reply:
Honestly, retries are also a common mechanism to handle such issues. You can configure the retry count to be high or set an eventual timeout. This approach will help your code scale easily in the future.